### PR TITLE
Dpp 1076 edit stop date page

### PR DESCRIPTION
**Pull request by davehowlett**

### CODE REVIEW

File: src/SFA.DAS.EAS.Web/Authentication/UserInfo.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿namespace SFA.DAS.EmployerCommitments.Web.Authentication
+{
+    public class UserInfo
```
Comment (chrisfoster76) [inline src/SFA.DAS.EAS.Web/Authentication/UserInfo.cs]: Why have we had to add this? OwinWrapper.GetClaimValue("sub") has been sufficient on other methods.

Comment (chrisfoster76) [inline src/SFA.DAS.EAS.Web/Authentication/UserInfo.cs]: Why have we had to add this? OwinWrapper.GetClaimValue("sub") has been sufficient on other methods.

Comment (davehowlett) [reply src/SFA.DAS.EAS.Web/Authentication/UserInfo.cs]: I've now removed this and used the existing. This had been added by Maulik to wrap up the OwinWrapper stuff, but he hadn't applied it elsewhere so it wasn't adding any benefit!

Comment (davehowlett) [reply src/SFA.DAS.EAS.Web/Authentication/UserInfo.cs]: I've now removed this and used the existing. This had been added by Maulik to wrap up the OwinWrapper stuff, but he hadn't applied it elsewhere so it wasn't adding any benefit!

File: src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs [hunk 2]
```diff
@@ -109,30 +109,30 @@ public class EmployerManageApprenticesController : BaseController
...
             var validationFailed = validatorResult != null && validatorResult.Count > 0;
 
+            var userId = OwinWrapper.GetClaimValue(@"sub");
+
             if (ModelState.IsValid && !validationFailed)
             {
-                var userInfo = GetExtractUserClaims();
-
-                var statusChangeModelWithNewStopDate = new ChangeStatusViewModel()
+                var statusChangeModelWithNewStopDate = new ChangeStatusViewModel
```
Comment (chrisfoster76) [inline src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs]: Should UpdateApprenticeshipStopDate use a ChangeStatusViewModel? This sounds like a VM for something else, and has properties like ChangeType= ChangeStatusType.Stop, which seems meaningless in this context.

File: src/SFA.DAS.EAS.Web/Orchestrators/EmployerManageApprenticeshipsOrchestrator.cs [hunk 3]
```diff
@@ -627,6 +630,43 @@ private bool CanChangeDateStepBeSkipped(ChangeStatusType changeType, GetApprenti
...
+
+                CheckApprenticeshipStateValidForChange(data.Apprenticeship);
+
+                await _mediator.SendAsync(new UpdateApprenticeshipStopDateCommand
+                {
+                    UserId = externalUserId,
+                    ApprenticeshipId = apprenticeshipId,
+                    EmployerAccountId = accountId,
+
+                    ChangeType = (Domain.Models.Apprenticeship.ChangeStatusType)model.ChangeType,
```
Comment (chrisfoster76) [inline src/SFA.DAS.EAS.Web/Orchestrators/EmployerManageApprenticeshipsOrchestrator.cs]: What is this ChangeType property? The only thing we can do with this command is change the stop date?

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommand.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using SFA.DAS.EmployerCommitments.Domain.Models.Apprenticeship;
+
+namespace SFA.DAS.EmployerCommitments.Application.Commands.UpdateApprenticeshipStopDate
+{
+    public sealed class UpdateApprenticeshipStopDateCommand : IAsyncRequest
+    {
+        public long ApprenticeshipId { get; set; }
+        public long EmployerAccountId { get; set; }
+        public string UserId { get; set; }
+        public ChangeStatusType ChangeType { get; set; }
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommand.cs]: See previous

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            };
+
+            await ValidateDateOfChange(command, validationResult);
+
+            await _commitmentsApi.PutApprenticeshipStopDate(command.EmployerAccountId, command.CommitmentId, command.ApprenticeshipId, stopDate);
+        }
+
+        private async Task ValidateDateOfChange(UpdateApprenticeshipStopDateCommand command, ValidationResult validationResult)
+        {
+            if (command.ChangeType == ChangeStatusType.Stop) // Only need to validate date for stop currently
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs]: See previous

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+            var stopDate = new ApprenticeshipStopDate
+            {
+                UserId = command.UserId,
+                LastUpdatedByInfo = new LastUpdateInfo { EmailAddress = command.UserEmailAddress, Name = command.UserDisplayName },
+                NewStopDate = command.DateOfChange
+            };
+
+            await ValidateDateOfChange(command, validationResult);
+
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs]: Should we also validate that the apprenticeship is actually stopped?

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                        var earliestDate = _academicYearDateProvider.CurrentAcademicYearStartDate.ToString("dd MM yyyy");
+
+                        validationResult.AddError(nameof(command.DateOfChange), $"The earliest date you can stop this apprentice is {earliestDate}");
+                        throw new InvalidRequestException(validationResult.ValidationDictionary);
+                    }
+                }
+            }
+        }
+
+        private static PaymentStatus DeterminePaymentStatusForChange(ChangeStatusType changeType)
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Commands/UpdateApprenticeshipStopDate/UpdateApprenticeshipStopDateCommandHandler.cs]: This is no longer a thing.

File: src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs (hunk 2 above)
Comment (chrisfoster76) [reply src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs]: Are we creating a viewmodel to pass to the orchestrator? Seems wrong. Some of those properties seem really odd anyway, like ChangeConfirmed=true?

Comment (Pauljgraham) [reply src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs]: Am changing to ChangeStopDateViewModel

File: src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs [hunk 8]
```diff
@@ -113,18 +113,18 @@ public class EmployerManageApprenticesController : BaseController
...
             if (ModelState.IsValid && !validationFailed)
             {
-                var statusChangeModelWithNewStopDate = new ChangeStatusViewModel
-                {
-                    ChangeType= ChangeStatusType.Stop,
-                    DateOfChange = model.NewStopDate,
-                    WhenToMakeChange = WhenToMakeChangeOptions.SpecificDate,
-                    ChangeConfirmed = true
-                };
+                //var statusChangeModelWithNewStopDate = new ChangeStatusViewModel
```
Comment (chrisfoster76) [inline src/SFA.DAS.EAS.Web/Controllers/EmployerManageApprenticesController.cs]: A commented block left over

**Comment by satya-sanikommu** [comment]
@Pauljgraham Please resolve the conflicts and merge master into it, so we will start testing

**Comment by satya-sanikommu** [comment]
Testing & Review completed. No regression issues. Ready to merge

**Pull request merged**